formats/mdm/MDM_2_4.ddl

```diff
@@ -0,0 +1,515 @@
+    major = UInt8
+    minor = UInt8
+
+-- See 6.1.2.1.5: ML2B interface ID
+def ML2B_Interface_id =
+  block
+    origination_id = BEUInt16
+    destination_id = BEUInt16
+
+-- Parse the message type followed by the number of messages field. This
+-- returns only the type since the number of messages is always implied
+-- by the type anyway.
+--
+-- See:
+-- * 6.1.2.1.6: MDM type
+-- * 6.1.2.1.7: Number of messages
+def MDM_Type (raw_ack_byte: uint 8) =
+  First
+    -- Parse the message type octet followed by the number of messages
+    -- octet, which has a fixed value corresponding to each message
+    -- type.
+    Type_VRT                  = @Match [0x0, 0x1, 0x0, 0x1]
+    Type_Time_Of_Day          = @Match [0x0, 0x3, 0x0, 0x1]
+    Type_Signal_Port_User_ID  = @Match [0x0, 0x4, 0x0, 0x1]
+    Type_Health_Status        = @Match [0x0, 0x5, 0x0, 0x1]
+    Type_Command              = @Match [0x0, 0x6, 0x0, 0x1]
+    Type_Switch_Group_User_ID = @Match [0x0, 0x7, 0x0, 0x1]
+
+    Type_Acknowledgement = block
+      @Match [0x0, 0x2, 0x0, 0x0]
+      -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+      --
+      -- In the Acknowledge messge type, the "ack" octet parsed in the
+      -- message header is treated as a response octet which takes one
+      -- of a number of meaningful values.
+      --
+      -- The specification states that all other possible values are
+      -- Reserved for Future Use; if they are encountered here, they
+      -- result in a parse failure.
+      response = raw_ack_byte as? MDM_Ack_Response_Value
+
+-- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification, Table
+-- 74
+bitdata MDM_Ack_Response_Value where
+  -- Acknowledged (A)
+  Resp_A = 0x01
+  -- Invalid Message (IM)
+  Resp_IM = 0x02
+  -- Invalid Function (IF)
+  Resp_IF = 0x04
+  -- Invalid Range (IR)
+  Resp_IR = 0x08
+  -- Resource Unavailable (RU)
+  Resp_RU = 0x10
+  -- Unexpected Sequence Number (US)
+  Resp_US = 0x20
+  -- Unauthorized User (UU)
+  Resp_UU = 0x40
+
+-- See 6.1.2.2, MDM Body
+def MDM_Body (ty: MDM_Type) =
+  case ty of
+    Type_VRT ->
+      -- See 6.3.1.2.1: VRT Message Format
+      -- VRT_Message
+      Fail "VRT message parsing not implemented"
+
+    -- See 6.3.2: MDM Acknowledgement Message (Type 2) Specification
+    -- Acknowledgement messages have no body.
+    Type_Acknowledgement info ->
+      ^ {| Body_Acknowledgement = { } |}
+
+    -- See 6.3.3: MDM Time of Day Message (Type 3) Specification
+    Type_Time_Of_Day ->
+      {| Body_Time_Of_Day = BEUInt32 |}
+
+    -- See 6.3.4: MDM Signal Port User ID Message (Type 4) Specification
+    Type_Signal_Port_User_ID ->
+      Parse_Body_Signal_Port_User_ID
+
+    -- See 6.3.5: MDM Health Message (Type 5) Specification
+    Type_Health_Status ->
+      block
+        let f1 = BEUInt32 as? Health_Status_Field_1
+        let f2 = BEUInt32 as? Health_Status_Field_2
+        ^ {| Body_Health_Status = { status_field_1 = f1, status_field_2 = f2 } |}
```

Comment:
Probably skip the ^
Also, I tend to avoid using anonymous types (e.g., the struct in the body of `Body_Health_Status`) as when generating code for the parser (e.g., C++) the resulting types don't have a stable name, so they are difficult to use from the application.

Reply:
The body of `Body_Health_Status` is actually the explicit struct type `Body_Health_Status_s`:
```
  def MDM_Body_u =
    union
      Body_VRT: {}
      Body_Acknowledgement: {}
      Body_Time_Of_Day: uint 32
      Body_Signal_Port_User_ID: Body_Signal_Port_User_ID_s
      Body_Health_Status: Body_Health_Status_s
      Body_Command: Body_Command_s
      Body_Switch_Group_User_ID: Body_Switch_Group_User_ID_s
```